File: src/Systems/Rust/adapter.rs
trait TemperatureReader {
    fn read_celsius(&self) -> i32;
}

struct LegacyFahrenheitSensor;

impl LegacyFahrenheitSensor {
    fn read_fahrenheit(&self) -> i32 {
        86
    }
}

struct FahrenheitSensorAdapter<'a> {
    sensor: &'a LegacyFahrenheitSensor,
}

impl TemperatureReader for FahrenheitSensorAdapter<'_> {
    fn read_celsius(&self) -> i32 {
        (self.sensor.read_fahrenheit() - 32) * 5 / 9
    }
}

fn main() {
    let legacy = LegacyFahrenheitSensor;
    let reader = FahrenheitSensorAdapter { sensor: &legacy };
    println!("legacy={}F", legacy.read_fahrenheit());
    println!("adapted={}C", reader.read_celsius());
}

File: src/Systems/Go/adapter.go
package main

import "fmt"

type TemperatureReader interface {
	ReadCelsius() int
}

type LegacyFahrenheitSensor struct{}

func (LegacyFahrenheitSensor) ReadFahrenheit() int {
	return 86
}

type FahrenheitSensorAdapter struct {
	sensor LegacyFahrenheitSensor
}

func (adapter FahrenheitSensorAdapter) ReadCelsius() int {
	return (adapter.sensor.ReadFahrenheit() - 32) * 5 / 9
}

func main() {
	legacy := LegacyFahrenheitSensor{}
	var reader TemperatureReader = FahrenheitSensorAdapter{sensor: legacy}
	fmt.Printf("legacy=%dF\n", legacy.ReadFahrenheit())
	fmt.Printf("adapted=%dC\n", reader.ReadCelsius())
}
